### DISCUSSION TITLE
Notes on new feature: Threaded Scheduler

### DISCUSSION THREAD

**Discussion by gabriels1234** [dsoftwareinc/django-rq-scheduler#96] [General]
Hi @cunla , how are you?
I am very happy the project is making a lot of progress.
Could you tell me a bit more on the Threaded Scheduler?
What is the motivation and what results/improvements did you see?
Thanks!

**Comment by cunla** [comment]
Hi @gabriels1234, all good.. it was a productive weekend :)

Overall my motivation is to reduce the dependency of this package on other packages since each dependency adds an additional risk of failure. I expect by the end of the year, this package will be dependent (for deployment, not for testing) on Django alone.

Regarding the threaded scheduler initiative:
- The `rq-scheduler` requirement was removed. However, the internal `rq` scheduler is not mature, so I created the `DjangoRQScheduler`, which inherits from `rq` scheduler (for now).
- I thought about triggering it via a django management command (something similar to `python manage.py rqscheduler`) but that makes it similar to depending on `rq-scheduler` => I figured better to have it running as a daemon.
- The implications are if you have multiple django processes, which you probably do on deployments (gunicorn, etc.), you have multiple schedulers as well. 
  - I verified that having multiple schedulers does not mean a job will be scheduled multiple times, there is a "lock" preventing multiple schedulers from checking the same queue. => There is a chance of having multiple schedulers scheduling at the same time but for different queues.
  - I haven't seen any drop in performance - but then again, we are not talking about millions of jobs. If you have thousands of scheduled jobs, the performance of responding to web requests will probably take a hit - in that scenario, having a separate process makes sense. I will create an issue for that.

**Reply by gabriels1234** [reply]
Interesting!

Did you mean to end up depending only on Django **and rq**?
Regarding multiple schedulers, currently there's no "**rq** way" of running just a scheduler (unless you bought on the idea of making the jobs be stored in the database, that would be awesome!) . 
That would necessarily mean be an additional worker enabled for all queues. I don't see benefit of it (different queues might require different containers, some bigger than the others) Also, some queues cannot have more than 1 worker since jobs might required to be handled sequentialy.

-(side point) I assume you agree that, if there are two workers, one on the queue default and the other on the queue low, there will be 2 schedulers running (which are the workers themselves), schedulers in the rq world are assigned to queues, the same queues the worker is enabled on. (correct me if I'm wrong)

- I found out in the projects I implemented 2023.2.1 that it works like a charm! And no more guessing what's going on with the scheduled jobs, they are all seen, the enqueue_at is now transparent (shown in the scheduled column) (also the Cron Jobs and repeatable jobs just work). 
   I put in place a Repeatable job to make all enabled jobs that are not scheduled to be re-scheduled (this was as you were adding the reschedule_all_jobs, which I left as well, but running every 15 minutes). I never had 1 single issue. It's as robust as it can get.
   So I don't feel there's a need to put-up the effort of making a new scheduler. (which as you mentioned, it's not well suited for prod environments). 
   
   Please let me know your thoughts, thanks!

**Comment by cunla** [comment]
You don't have to use it, you can opt-out by having `SCHEDULER_THREAD=False` in your settings.

**Reply by gabriels1234** [reply]
@cunla I was checking rq's source code, and the with_scheduler logic is exactly that, not that te worker also does scheduling before/after, but an actual thread running the scheduler. What's the difference with yours? what does " the internal rq scheduler is not mature". I know that was said by their own authors. But we are now probably a few years after that.

Any issues you faced with 2023.2.1 might be important to understand this move to .3.

**Reply by cunla** [reply]
As mentioned before, you can opt out by having `SCHEDULER_THREAD=False` in your settings.

I did not face any issues using 2023.2.1. 
However, 2023.3 is a serious refactoring with the runtime parameters and that could not be supported easily with the previous scheduler.

**Reply by gabriels1234** [reply]
My question is: What's the motivation behind? Also, why make SCHEDULER_THREAD default to True in its first appearance for production deployments is a perhaps risky step, but perhaps it's worth doing it.
consider this code:

```
        self.thread = threading.Thread(
        target=run, args=(self,), name='Scheduler', daemon=True)
```
vs

```
        self._process = Process(target=run, args=(self,), name='Scheduler')
        self._process.start()
```

whereas Process is:
<img width="540" alt="image" src="https://user-images.githubusercontent.com/96071233/220975042-1502a009-8443-4b51-912d-5bfbae25c5d5.png">

So it would Seem running it with SCHEDULER_THREAD=True is almost the same, little risk of breaking things.

Basically, the only true fix is for those who don't run the worker --with-scheduler, Django-RQ-Scheduler will make sure this is the case.
Any other benefit? is Threading.Thread better than the Process class?

**Reply by gabriels1234** [reply]
More digging: I didn't realize earlier that this is actually Multiprocessing vs Multithreading.
Don't you think this is an analysis/change to be made in RQ itself? Should I bring this topic in RQ's issues?

**Reply by cunla** [reply]
The analysis has been done multiple times ([first result on google](https://stackoverflow.com/a/3046201/1056460)).

I prefer to enable users to choose here.

**Reply by gabriels1234** [reply]
I understand. Thanks!

**Comment by gabriels1234** [comment]
@cunla seeing it in action, this message could be confusing for users:
"Internal scheduler thread status: No internal scheduler ([DjangoRQ info])"
This means that the Threaded scheduler is not enabled.
However, given that the Multiprocessing Scheduler is native to RQ, (the only dependency planned in the future) I believe some accommodations should be made so it's clear what's going on to the end user.
Moreover, it's worth noting that a Threaded scheduler could be sharing cpu with the worker, whereas the other could make use of another cpu core (that, and the memory-leak proof of the Multiprocessing approach). I'm just saying DjangoRQScheduler should also be friendly to RQ's (native --with-scheduler) Scheduler.

**Reply by cunla** [reply]
How would you change the wording? I thought "internal scheduler thread" is a good distinguish but I am open to suggestions while removing the django-rq dependency altogether.

**Reply by gabriels1234** [reply]
I think we have three issues that are interesting to discuss.

How to make Dj-RQ-S (Django-RQ-Scheduler) evolve from its current shape to its new shape without taxing current users (that might upgrade and upgrade and "upgrade-into-oblivion").

Given that Django-rq is an official sub-project of RQ, it shouldn't be perceived as one going to war against django-rq.

Here I propose Dj-RQ-S to operate based on awareness. Anything that can coexist (namely: Dj-RQ-S and Django-RQ), should coexist (i.e not break Django-RQ).
In the Ideal world, people would only add RQ and Dj-RQ-S to the requirements and Dj-RQ-S would not install Django-rq.

What about the Dj-RQ-S Scheduler?

Here I propose need-base activation.
Adding two settings:
```
# This will define if it needs to instantiate a scheduler. Querying Redis for the scheduler_lock
# if value is OFF, never start the Dj-RQ-S Scheduler, if AUTO, a (own) scheduler will start only if no other scheduler is present. (this is worth checking any time possible)
DJANGO_RQ_SCHEDULER_SCHEDULER = 'AUTO' (DEFAULT) / 'OFF'
# this is to decide vs Thread and process, False would start Dj-RQ-S Scheduler in multiprocessing mode.
SCHEDULER_THREAD = True (Default) /False
```

The message: `"Internal scheduler thread status: No internal scheduler ([DjangoRQ info])"`Makes sense only if there is no scheduler active (easy to check in rq with the locks) AND DJANGO_RQ_SCHEDULER_SCHEDULER is set to OFF.
This is just saying "DJANGO_RQ_SCHEDULER" won't work without that....

Last but not least, if django_rq's dashboard is to be replaced, make one that auto-refreshed every couple seconds...

Thanks!!

**Reply by cunla** [reply]
I am not planning to go to war with RQ and django-rq :) simply to remove the dependency of this package on them. I aim to have a separation, i.e., reach a situation where you can have RQ workers with schedulers, but they won't be used by django-rq-schduler, and vice versa.

I am not planning to release breaking API more than twice a year, and when I do so, I will give proper notice.

For your purposes: I suggest you override the admin template on your app (simply remove the message there).

**Reply by gabriels1234** [reply]
Great!.
"i.e. reach a situation where you can have RQ workers with schedulers, but they won't be used by django-rq-scheduler, and vice versa."
I assume that you plan to make this change "optional", since it's very common for workers to run on different containers Some with more memory, some with less, but definitely separated from the main web/django container. Certainly, RQ is used for more than scheduled/cron jobs. but there's also a benefit of having a "default" queue handled by an independent worker-scheduler.
